File: Sources/RefId/Parse.swift
// SPDX-License-Identifier: Apache-2.0
//
// Decomposes a `ref:` identifier string against the loaded spec. Order, and the precedence it
// produces: grammar → version → state and fragment decomposition, including the repeated-key policy
// → (unsupported version stops here, decomposed and unvalidated) → the two sides must not trade keys
// → qualifier values against their forms → refinement values against their patterns and range
// constraints → dispatch: unknown type is `uncovered`, else the locator goes to its validator.

import Foundation

private enum Decomposed<T> {
    case ok(T)
    case failed(part: String)
}

private struct Parser {
    let spec: Spec
    let grammar: Grammar

    func malformed(_ input: String, _ failedPart: String, _ head: ParseResult?) throws -> ParseResult {
        var result = head ?? ParseResult(input: input, status: "", version: spec.int("version", "default"), explicitVersion: false, type: "", locator: "", qualifiers: [], fragment: nil)
        result.input = input
        result.status = try spec.status("malformed")
        result.delegated = nil
        result.canonical = nil
        result.nested = nil
        result.part = try spec.part(failedPart)
        return result
    }

    func group(_ match: Regex<AnyRegexOutput>.Match, _ name: String) -> String? {
        guard let element = match[name] as AnyRegexOutput.Element?, let substring = element.substring else { return nil }
        return String(substring)
    }

    /// Splits `key=value` segments with the given pair grammar; a repeated key is malformed at that key.
    func pairs(_ regex: Regex<AnyRegexOutput>, _ segments: [String], _ failedPart: String) -> Decomposed<[Pair]> {
        var out: [Pair] = []
        var seen = Set<String>()
        for segment in segments {
            guard let match = try? regex.wholeMatch(in: segment), let key = group(match, "key") else { return .failed(part: failedPart) }
            if seen.contains(key) { return .failed(part: key) }
            seen.insert(key)
            out.append(Pair(key, group(match, "value") ?? ""))
        }
        return .ok(out)
    }

    func state(_ raw: String) -> Decomposed<[Pair]> {
        pairs(grammar.statePair, raw.components(separatedBy: spec.string("grammar", "state", "separator")), "state")
    }

    func fragment(_ raw: String) -> Decomposed<Fragment> {
        let segments = raw.components(separatedBy: spec.string("grammar", "fragment", "separator"))
        let path = segments.first ?? ""
        if path.isEmpty { return .failed(part: "fragment") }
        switch pairs(grammar.fragmentPair, Array(segments.dropFirst()), "fragment") {
        case .failed(let part): return .failed(part: part)
        case .ok(let refinements): return .ok(Fragment(path: path, refinements: refinements))
        }
    }

    /// A qualifier value that nests an identifier: strictly encoded, decoded with the form's table, parsed one level down.
    func tryNested(_ form: [String: Any], _ value: String, depth: Int) throws -> String? {
        let maxDepth = (form["depth"] as? NSNumber)?.intValue ?? 0
        guard depth < maxDepth, value.hasPrefix(schemePrefix(spec)) else { return nil }
        let table = Encoding.table(spec, form: form)
        guard Encoding.strictlyEncoded(value, table: table) else { return nil }
        let decoded = Encoding.decode(value, table: table)
        let inner = try parse(decoded, depth: depth + 1)
        return inner.status == (try spec.status("malformed")) ? nil : decoded
    }

    func matchForms(_ names: [String], _ value: String, depth: Int) throws -> (matches: Bool, nested: String?) {
        for name in names {
            guard let form = spec.dictionary("forms")[name] as? [String: Any] else { continue }
            if (form["nested"] as? NSNumber)?.boolValue == true {
                if let nested = try tryNested(form, value, depth: depth) { return (true, nested) }
                continue
            }
            if let pattern = form["pattern"] as? String, try grammar.pattern(pattern).matches(value) { return (true, nil) }
        }
        return (false, nil)
    }

    func parse(_ input: String, depth: Int) throws -> ParseResult {
        guard let match = try? grammar.top.wholeMatch(in: input) else { return try malformed(input, "grammar", nil) }
        let versionText = group(match, "version")
        let type = group(match, "type") ?? ""
        let locator = group(match, "locator") ?? ""
        let stateRaw = group(match, "state")
        let fragmentRaw = group(match, "fragment")

        let explicitVersion = versionText != nil
        let version = versionText.flatMap { Int($0) } ?? spec.int("version", "default")
        var head = ParseResult(input: input, status: "", version: version, explicitVersion: explicitVersion, versionText: versionText, type: type, locator: locator, qualifiers: [], fragment: nil)

        if let stateRaw {
            switch state(stateRaw) {
            case .failed(let part): return try malformed(input, part, head)
            case .ok(let qualifiers): head.qualifiers = qualifiers
            }
        }
        if let fragmentRaw {
            switch fragment(fragmentRaw) {
            case .failed(let part): return try malformed(input, part, head)
            case .ok(let fragment): head.fragment = fragment
            }
        }

        var base = head
        base.status = try spec.status("ok")
        base.delegated = Validators.delegatedString(spec, type: type, locator: locator)

        if !spec.ints("version", "supported").contains(version) {
            base.status = try spec.status("unsupported")
            return base
        }

        let refinementsTable = spec.dictionary("refinements")
        let qualifiersTable = spec.dictionary("qualifiers")
        for pair in head.qualifiers where refinementsTable[pair.key] != nil {
            return try malformed(input, pair.key, head)
        }
        if let fragment = head.fragment {
            for pair in fragment.refinements where qualifiersTable[pair.key] != nil {
                return try malformed(input, pair.key, head)
            }
        }

        var nested: [String: String] = [:]
        for pair in head.qualifiers {
            guard let declared = qualifiersTable[pair.key] as? [String: Any] else { continue } // unknownKey: carry-through
            let forms = declared["forms"] as? [String] ?? []
            let result = try matchForms(forms, pair.value, depth: depth)
            if !result.matches { return try malformed(input, pair.key, head) }
            if let value = result.nested { nested[pair.key] = value }
        }
        if !nested.isEmpty { base.nested = nested }

        if let fragment = head.fragment {
            for pair in fragment.refinements {
                guard let declared = refinementsTable[pair.key] as? [String: Any] else { continue } // unknownRefinement: carry-through
                let pattern = declared["pattern"] as? String ?? ""
                if try !grammar.pattern(pattern).matches(pair.value) || !Validators.rangeHolds(declared, value: pair.value) {
                    return try malformed(input, pair.key, head)
                }
            }
        }

        guard let entry = spec.dictionary("dispatch")[type] as? [String: Any], let delegated = base.delegated else {
            base.status = spec.unknownType
            return base
        }
        guard let validation = try Validators.validateLocator(spec, entry: entry, delegated: delegated) else {
            base.status = spec.unknownType
            return base
        }
        if !validation.ok { return try malformed(input, "locator", head) }
        base.canonical = validation.canonical
        return base
    }
}

/// Parses a `ref:` identifier string against the embedded spec. Never throws for an identifier problem;
/// it throws only when the embedded specification itself cannot be honoured.
public func parse(_ input: String) throws -> ParseResult {
    let spec = try loadSpec()
    try Validators.assertImplemented(spec)
    return try Parser(spec: spec, grammar: Grammar.of(spec)).parse(input, depth: 0)
}

File: crates/ref-id/src/parse.rs
// SPDX-License-Identifier: Apache-2.0
//
// Decomposes a `ref:` identifier string against the loaded spec. Order, and the precedence it
// produces: grammar → version → state and fragment decomposition, including the repeated-key policy
// → (unsupported version stops here, decomposed and unvalidated) → the two sides must not trade keys
// → qualifier values against their forms → refinement values against their patterns and range
// constraints → dispatch: unknown type is `uncovered`, else the locator goes to its validator.

use crate::encoding::{decode, strictly_encoded, table_for};
use crate::grammar::{scheme_prefix, Grammar};
use crate::spec::{load_spec, Spec};
use crate::types::{Fragment, Pair, ParseResult, RefIdError};
use crate::validators::{assert_implemented, delegated_string, range_holds, validate_locator};
use regex::Regex;
use serde_json::Value;
use std::collections::HashSet;

enum Decomposed<T> {
    Ok(T),
    Failed(String),
}

struct Parser<'a> {
    spec: &'a Spec,
    grammar: &'a Grammar,
}

impl<'a> Parser<'a> {
    fn malformed(&self, input: &str, part: &str, head: Option<&ParseResult>) -> Result<ParseResult, RefIdError> {
        let mut result = head.cloned().unwrap_or(ParseResult {
            input: String::new(),
            status: String::new(),
            version: self.spec.int(&["version", "default"]),
            explicit_version: false,
            version_text: None,
            r#type: String::new(),
            locator: String::new(),
            delegated: None,
            canonical: None,
            qualifiers: Vec::new(),
            nested: None,
            fragment: None,
            part: None,
        });
        result.input = input.to_string();
        result.status = self.spec.status("malformed")?;
        result.delegated = None;
        result.canonical = None;
        result.nested = None;
        result.part = Some(self.spec.part(part)?);
        Ok(result)
    }

    /// Splits `key=value` segments with the given pair grammar; a repeated key is malformed at that key.
    fn pairs(&self, regex: &Regex, segments: &[&str], failed_part: &str) -> Decomposed<Vec<Pair>> {
        let mut out = Vec::new();
        let mut seen = HashSet::new();
        for segment in segments {
            let Some(captures) = regex.captures(segment) else { return Decomposed::Failed(failed_part.to_string()) };
            let key = captures.name("key").map(|m| m.as_str()).unwrap_or("").to_string();
            if !seen.insert(key.clone()) {
                return Decomposed::Failed(key);
            }
            out.push(Pair::new(key, captures.name("value").map(|m| m.as_str()).unwrap_or("")));
        }
        Decomposed::Ok(out)
    }

    fn state(&self, raw: &str) -> Decomposed<Vec<Pair>> {
        let segments: Vec<&str> = raw.split(self.spec.string(&["grammar", "state", "separator"])).collect();
        self.pairs(&self.grammar.state_pair, &segments, "state")
    }

    fn fragment(&self, raw: &str) -> Decomposed<Fragment> {
        let segments: Vec<&str> = raw.split(self.spec.string(&["grammar", "fragment", "separator"])).collect();
        let path = segments.first().copied().unwrap_or("");
        if path.is_empty() {
            return Decomposed::Failed("fragment".into());
        }
        match self.pairs(&self.grammar.fragment_pair, &segments[1..], "fragment") {
            Decomposed::Failed(part) => Decomposed::Failed(part),
            Decomposed::Ok(refinements) => Decomposed::Ok(Fragment { path: path.to_string(), refinements }),
        }
    }

    /// A qualifier value that nests an identifier: strictly encoded, decoded with the form's table, parsed one level down.
    fn try_nested(&self, form: &Value, value: &str, depth: i64) -> Result<Option<String>, RefIdError> {
        let max_depth = form.get("depth").and_then(Value::as_i64).unwrap_or(0);
        if depth >= max_depth || !value.starts_with(&scheme_prefix(self.spec)) {
            return Ok(None);
        }
        let table = table_for(self.spec, form);
        if !strictly_encoded(value, &table) {
            return Ok(None);
        }
        let decoded = decode(value, &table);
        let inner = self.parse(&decoded, depth + 1)?;
        Ok(if inner.status == self.spec.status("malformed")? { None } else { Some(decoded) })
    }

    fn match_forms(&self, names: &[String], value: &str, depth: i64) -> Result<(bool, Option<String>), RefIdError> {
        let forms = self.spec.object(&["forms"]);
        for name in names {
            let Some(form) = forms.and_then(|m| m.get(name)) else { continue };
            if form.get("nested").and_then(Value::as_bool) == Some(true) {
                if let Some(nested) = self.try_nested(form, value, depth)? {
                    return Ok((true, Some(nested)));
                }
                continue;
            }
            if let Some(pattern) = form.get("pattern").and_then(Value::as_str) {
                if self.grammar.matches(self.spec, pattern, value)? {
                    return Ok((true, None));
                }
            }
        }
        Ok((false, None))
    }

    fn parse(&self, input: &str, depth: i64) -> Result<ParseResult, RefIdError> {
        let Some(captures) = self.grammar.top.captures(input) else { return self.malformed(input, "grammar", None) };
        let group = |name: &str| captures.name(name).map(|m| m.as_str().to_string());
        let version_text = group("version");
        let r#type = group("type").unwrap_or_default();
        let locator = group("locator").unwrap_or_default();
        let explicit_version = version_text.is_some();
        let version = version_text.as_deref().and_then(|v| v.parse::<i64>().ok()).unwrap_or_else(|| self.spec.int(&["version", "default"]));

        let mut head = ParseResult {
            input: input.to_string(),
            status: String::new(),
            version,
            explicit_version,
            version_text: version_text.clone(),
            r#type: r#type.clone(),
            locator: locator.clone(),
            delegated: None,
            canonical: None,
            qualifiers: Vec::new(),
            nested: None,
            fragment: None,
            part: None,
        };

        if let Some(state) = group("state") {
            match self.state(&state) {
                Decomposed::Failed(part) => return self.malformed(input, &part, Some(&head)),
                Decomposed::Ok(qualifiers) => head.qualifiers = qualifiers,
            }
        }
        if let Some(fragment) = group("fragment") {
            match self.fragment(&fragment) {
                Decomposed::Failed(part) => return self.malformed(input, &part, Some(&head)),
                Decomposed::Ok(fragment) => head.fragment = Some(fragment),
            }
        }

        let mut base = head.clone();
        base.status = self.spec.status("ok")?;
        base.delegated = delegated_string(self.spec, &r#type, &locator);

        if !self.spec.ints(&["version", "supported"]).contains(&version) {
            base.status = self.spec.status("unsupported")?;
            return Ok(base);
        }

        let refinements_table = self.spec.object(&["refinements"]);
        let qualifiers_table = self.spec.object(&["qualifiers"]);
        for pair in &head.qualifiers {
            if refinements_table.is_some_and(|m| m.contains_key(&pair.key)) {
                return self.malformed(input, &pair.key, Some(&head));
            }
        }
        if let Some(fragment) = &head.fragment {
            for pair in &fragment.refinements {
                if qualifiers_table.is_some_and(|m| m.contains_key(&pair.key)) {
                    return self.malformed(input, &pair.key, Some(&head));
                }
            }
        }

        let mut nested: Vec<(String, String)> = Vec::new();
        for pair in &head.qualifiers {
            let Some(declared) = qualifiers_table.and_then(|m| m.get(&pair.key)) else { continue }; // unknownKey: carry-through
            let forms: Vec<String> = declared.get("forms").and_then(Value::as_array).map(|a| a.iter().filter_map(Value::as_str).map(String::from).collect()).unwrap_or_default();
            let (matches, inner) = self.match_forms(&forms, &pair.value, depth)?;
            if !matches {
                return self.malformed(input, &pair.key, Some(&head));
            }
            if let Some(inner) = inner {
                nested.push((pair.key.clone(), inner));
            }
        }
        if !nested.is_empty() {
            base.nested = Some(nested);
        }

        if let Some(fragment) = &head.fragment {
            for pair in &fragment.refinements {
                let Some(declared) = refinements_table.and_then(|m| m.get(&pair.key)) else { continue }; // unknownRefinement: carry-through
                let pattern = declared.get("pattern").and_then(Value::as_str).unwrap_or("");
                if !self.grammar.matches(self.spec, pattern, &pair.value)? || !range_holds(declared, &pair.value) {
                    return self.malformed(input, &pair.key, Some(&head));
                }
            }
        }

        let entry = self.spec.object(&["dispatch"]).and_then(|m| m.get(&r#type));
        let (Some(entry), Some(delegated)) = (entry, base.delegated.clone()) else {
            base.status = self.spec.unknown_type().to_string();
            return Ok(base);
        };
        let Some(validation) = validate_locator(self.spec, entry, &delegated)? else {
            base.status = self.spec.unknown_type().to_string();
            return Ok(base);
        };
        if !validation.ok {
            return self.malformed(input, "locator", Some(&head));
        }
        base.canonical = validation.canonical;
        Ok(base)
    }
}

/// Parses a `ref:` identifier string against the embedded spec. Never errs for an identifier problem;
/// it errs only when the embedded specification itself cannot be honoured.
pub fn parse(input: &str) -> Result<ParseResult, RefIdError> {
    let spec = load_spec()?;
    assert_implemented(spec)?;
    Parser { spec, grammar: Grammar::of(spec)? }.parse(input, 0)
}
